applications: pelion_client: Add pelion client

pdunaj opened the pull request.

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-4365
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

pdunaj commented:
Fixing checkpatch issues. Removing timer tests as there is no chance we fix checkpatch fast.

pdunaj commented:
Fix more checkpatch issues.

pdunaj commented:
Hi @rlubos , @carlescufi , I fixed all fixable checkpatch issues. Can you have a look and give waiver for the remaining ones?

pdunaj commented:
Hi @jumi-nordic , can you try to run pelion tests against this commit?

pdunaj commented:
Hi @rakons , @emob-nordic , can one of you fetch the PR's branch and diff it with our dev repo. During the move I had to modify few files. The changes are:
- manifest
- removal of timer tests
- fixes for checkpack (especially look at events and power manager, except that it should be minor changes).

I have checked application of nrf91 and it works the same as on dev branch.

emob-nordic commented:
> Hi @rakons , @emob-nordic , can one of you fetch the PR's branch and diff it with our dev repo. During the move I had to modify few files. The changes are:
> 
> * manifest
> * removal of timer tests
> * fixes for checkpack (especially look at events and power manager, except that it should be minor changes).
> 
> I have checked application of nrf91 and it works the same as on dev branch.

I will check that.

tejlmand commented on applications/pelion_client/CMakeLists.txt:
```diff
@@ -0,0 +1,102 @@
+  set(CMAKE_BUILD_TYPE ZDebug)
+endif()
+
+if (CONFIG_NET_L2_OPENTHREAD)
+  set(OT_CONFIG "${CMAKE_CURRENT_SOURCE_DIR}/configuration/common/ot_config.h")
+endif()
```
This won't work, as you don't know if `CONFIG_NET_L2_OPENTHREAD=y` or not.

pdunaj replied:
What is the correct way to check this?

rakons replied:
#if IS_ENABLED(CONFIG_NET_L2_OPENTHREAD)

pdunaj replied:
but this is CMakeLists.txt. The code here is actually the same as everywhere else. I think it was working fine. But maybe this option is somehow different... Let's double check.

rakons replied:
Right. So just if should make the job.

See the documentation:

```
 if(<constant>)
 
    True if the constant is 1, ON, YES, TRUE, Y, or a non-zero number. False if the constant is 0, OFF, NO, FALSE, N, IGNORE, NOTFOUND, the empty string, or ends in the suffix -NOTFOUND. Named boolean constants are case-insensitive. If the argument is not one of these specific constants, it is treated as a variable or string and 
the following signature is used.
```

pdunaj replied:
But in this case what is wrong here? @tejlmand

tejlmand replied:
The problem is that if user specify `CONFIG_NET_L2_OPENTHREAD=y` in a config file, this value will be undefined at this point in time, because `find_package(Zephyr)` has not been executed.

pdunaj replied:
right - this is done before the boilerplate is executed.
Why do we do it anyway @rakons ? It seems it is indeed not set so I guess we can safely remove it.

pdunaj replied:
I see that there is another build define set `OPENTHREAD_PROJECT_CORE_CONFIG_FILE` that pass info about config file.
I am removing it.

pdunaj replied:
done

tejlmand commented on applications/pelion_client/CMakeLists.txt:
```diff
@@ -0,0 +1,102 @@
+  set(DTC_OVERLAY_FILE "${CMAKE_CURRENT_SOURCE_DIR}/configuration/${BOARD}/dts.overlay")
+endif()
+
+set(mcuboot_CONF_FILE
+  ${CMAKE_CURRENT_SOURCE_DIR}/configuration/${BOARD}/mcuboot_${CMAKE_BUILD_TYPE}.conf
+  )
```
This is not needed if using the `prj_<build>.conf` approach, as you can then use out common child image feature:
http://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_multi_image.html#image-specific-variables

pdunaj replied:
prj is no go for us.

tejlmand replied:
accepted as-is for current time.
Let's continue discussions on way forward, and adopt all apps to the final decision.

tejlmand commented on applications/pelion_client/CMakeLists.txt:
```diff
@@ -0,0 +1,102 @@
+  ${CMAKE_CURRENT_SOURCE_DIR}/configuration/${BOARD}/mcuboot_${CMAKE_BUILD_TYPE}.conf
+  )
+
+if (EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/configuration/${BOARD}/spm_${CMAKE_BUILD_TYPE}.conf)
+  set(spm_CONF_FILE
+    ${CMAKE_CURRENT_SOURCE_DIR}/configuration/${BOARD}/spm_${CMAKE_BUILD_TYPE}.conf
+    )
+endif()
```
Same child image conf comment.

tejlmand replied:
accepted as-is for current time.
Let's continue discussions on way forward, and adopt all apps to the final decision.

tejlmand commented on applications/pelion_client/CMakeLists.txt:
```diff
@@ -0,0 +1,102 @@
+  ${CMAKE_CURRENT_SOURCE_DIR}/configuration/${BOARD}/pm_static_${CMAKE_BUILD_TYPE}.yml
+  )
+
+# Define configuration files.
+set(CONF_FILE "configuration/${BOARD}/app_${CMAKE_BUILD_TYPE}.conf")
```
can also be removed.

tejlmand replied:
accepted as-is for current time.
Let's continue discussions on way forward, and adopt all apps to the final decision.

tejlmand commented on applications/pelion_client/CMakeLists.txt:
```diff
@@ -0,0 +1,102 @@
+
+if(CONFIG_PELION_MBEDTLS)
+  # Copy configuration file to include directory visible for all libraries.
+  # Needed for the file to be visible by Zephyr's mbedtls library.
```
Are you not using nrf_security ?
And why not.

pdunaj replied:
We do by default. There is a second configuration that use zephyr mbedtls but we see some problems with it so we will have to either fix it or remove it.

tejlmand replied:
Ok, thanks for clarification.

tejlmand commented on applications/pelion_client/CMakeLists.txt:
```diff
@@ -0,0 +1,102 @@
+cmake_minimum_required(VERSION 3.13.1)
+
+
+if (NOT CMAKE_BUILD_TYPE)
+  set(CMAKE_BUILD_TYPE ZDebug)
+endif()
```
Please use the Zephyr build types instead.
https://docs.zephyrproject.org/latest/guides/kconfig/setting.html#the-initial-configuration

This means you can create:
- `prj.conf`: The default config file. Corresponding to what you have in `ZDebug.conf` today.
- `prj_mbedtls.conf`: The mbedtls configuration. Corresponding to what you have in `ZDebugMbedTLS.conf` today..
- `prj_release.conf`: This could be a release config.

The advantage of this approach is that it follows Zephyr build system, and that you will get corresponding board adjustment specified, for example building:
- `west build -bnrf52840dk_nrf52840` would include `boards/nrf52840dk_nrf52840.conf`
- `west build -bnrf52840dk_nrf52840 -- -DCONF_FILE=prj_mbedtls.conf` would include `boards/nrf52840dk_nrf52840_mbedtls.conf` instead.

pdunaj replied:
This won't work for us. I can discuss this offline in detail but in short this architecture allows to have separate caf configuration files for various boards.
Similarly as is at the moment done for applications/machine_learning and applications/nrf_desktop. 
If you find it a blocker it will be a serious issue that will impact the project.

tejlmand replied:
accepted as-is for current time.
Let's continue discussions on way forward, and adopt all apps to the final decision.

tejlmand reviewed (changes requested):
Please use the existing `prj_<build>.conf` infrastructure that exists today in Zephyr together with the child image conf infrastructure in NCS.

rakons reviewed (approved):
Nothing that is a real showstopper relative to our perlion repository. I am missing tests - cannot we leave unit tests here? They should not require any specialized configuration.

pdunaj commented:
> Nothing that is a real showstopper relative to our perlion repository. I am missing tests - cannot we leave unit tests here? They should not require any specialized configuration.

I would like to add them but these had a serious checkpatch issues. If we add them later on as a subsequent PR nothing will happen. Pelion PAL won't change at the moment (not at the area that is being tested anyway).

rlubos commented on applications/pelion_client/CMakeLists.txt:
```diff
@@ -0,0 +1,102 @@
+#
+# Copyright (c) 2020 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
```
Please verify the license headers in this PR, a lot of files still use the old license ref, `LicenseRef-Nordic-5-Clause` should be used instead.

pdunaj replied:
fixed

rlubos commented on applications/pelion_client/configuration/common/update_default_resources.c:
```diff
@@ -0,0 +1,39 @@
+// ----------------------------------------------------------------------------
```
Why doesn't checkpatch complain about `//` comments is a mystery to me.

pdunaj replied:
it's not perfect... anyway this file is a stub that is substituted by real file generated from tools. I missed it and will clean it up.

pdunaj replied:
fixed

rlubos commented on applications/pelion_client/src/modules/led_state.c:
```diff
@@ -0,0 +1,107 @@
+		return false;
+	}
+
+	/* If event is unhandled, unsubscribe. */
```
I'm a bit confused here, how does the assert relate to this comment?

pdunaj replied:
And you are rightfully confused. This is one of this things that get copy pasted so long that become invisible. The comment has little sense. It should be "Event is unhandled - subscription is missing." or something similar.

pdunaj replied:
Will fix it.

pdunaj replied:
Actually... this is not wrong. It can be just my poor English.
What is meant here it that: If event is unhandled, subscription is not needed so... unsubscribe".

pdunaj replied:
changed

rlubos commented on applications/pelion_client/src/modules/net_ot.cpp:
```diff
@@ -0,0 +1,213 @@
+	while (otThreadGetNextNeighborInfo(instance, &iterator, &info) == OT_ERROR_NONE) {
+		char addr_str[ARRAY_SIZE(info.mExtAddress.m8) * 2 + 1];
+
+		format_address(addr_str, sizeof(addr_str), info.mExtAddress.m8,
```
Not a big deal, but there's a standard `inet_ntop()` function for binary->string conversion. Unless you _really_ prefer a custom IPv6 string format (w/o colons).

pdunaj replied:
I prefer to use standard display method as long as I can use it, thanks.

pdunaj replied:
I will have to know the inet type here. I don't know it it will always be ipv6. Let's leave it as is.

rlubos commented on applications/pelion_client/src/modules/net_ot.cpp:
```diff
@@ -0,0 +1,213 @@
+/*
```
Why is this a cpp file? OT exposes C API.

pdunaj replied:
Don't know. Probably nobody checked it works for C and since OT lib is in C++ the file appeared as cpp. Will try to change the extension.

pdunaj replied:
nope - it fails as we refer to ot_context. so no change here

rlubos replied:
That's a bit weird ot_context is an issue here - it's still C.

pdunaj replied:
That's the static_cast.... will change and retry

pdunaj replied:
Fixed

rlubos commented:
I'm not sure how to fix the checkpatch issues to be honest (I assume those constants are used somewhere insde the pelion lib?). Is there a way to make checkpatch ignore parts of code other than adding the file to the exclude list? @PerMac @carlescufi

pdunaj commented:
> I'm not sure how to fix the checkpatch issues to be honest (I assume those constants are used somewhere insde the pelion lib?). Is there a way to make checkpatch ignore parts of code other than adding the file to the exclude list? @PerMac @carlescufi

Indeed Pelion is accessing this global variables from its internals.

tejlmand commented on lib/pelion/Kconfig:
```diff
@@ -0,0 +1,113 @@
+	default 4096
+
+config PELION_NRF_SECURITY
+	bool "Enable nrf_security module for OpenThread"
```
This has the flaw that if user just run:
```
$ cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -Bbuild .
$ ninja -Cbuild menuconfig
# Inside menuconfig, select `PELION_MBEDTLS=y`
# Save and exit
$ ninja -Cbuild
ninja: Entering directory `build'
[0/1] Re-running CMake...
Including boilerplate (Zephyr base (cached)): /projects/github/ncs/zephyr/cmake/app/boilerplate.cmake
-- Application: /projects/github/ncs/nrf/applications/pelion_client
-- Zephyr version: 2.5.99 (/projects/github/ncs/zephyr)
-- Found west (found suitable version "0.10.99", minimum required is "0.7.1")
-- Board: nrf52840dk_nrf52840
-- Cache files will be written to: /home/tora/.cache/zephyr
-- Found toolchain: gnuarmemb (/opt/gcc-arm-none-eabi-7-2018-q2-update)
-- Found BOARD.dts: /projects/github/ncs/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840.dts
-- Found devicetree overlay: /projects/github/ncs/nrf/applications/pelion_client/configuration/nrf52840dk_nrf52840/dts.overlay
-- Generated zephyr.dts: /projects/github/ncs/nrf/applications/pelion_client/build/zephyr/zephyr.dts
-- Generated devicetree_unfixed.h: /projects/github/ncs/nrf/applications/pelion_client/build/zephyr/include/generated/devicetree_unfixed.h
-- Generated device_extern.h: /projects/github/ncs/nrf/applications/pelion_client/build/zephyr/include/generated/device_extern.h

warning: MBEDTLS_USER_CONFIG_ENABLE (defined at modules/Kconfig.tls-generic:334) has direct dependencies MBEDTLS_BUILTIN && MBEDTLS_CFG_FILE = "config-tls-generic.h" && MBEDTLS with value n, but is currently being y-selected by the following symbols:
 - PELION_MBEDTLS (defined at /projects/github/ncs/nrf/lib/pelion/Kconfig:68), with value y, direct dependencies PELION_SOURCES (value: y), and select condition PELION_SOURCES (value: y)

warning: MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED (defined at modules/Kconfig.tls-generic:86) has direct dependencies MBEDTLS_BUILTIN && MBEDTLS_CFG_FILE = "config-tls-generic.h" && MBEDTLS with value n, but is currently being y-selected by the following symbols:
 - PELION_MBEDTLS (defined at /projects/github/ncs/nrf/lib/pelion/Kconfig:68), with value y, direct dependencies PELION_SOURCES (value: y), and select condition PELION_SOURCES (value: y)

warning: MBEDTLS_CIPHER_AES_ENABLED (defined at modules/Kconfig.tls-generic:189) has direct dependencies MBEDTLS_BUILTIN && MBEDTLS_CFG_FILE = "config-tls-generic.h" && MBEDTLS with value n, but is currently being y-selected by the following symbols:
 - PELION_MBEDTLS (defined at /projects/github/ncs/nrf/lib/pelion/Kconfig:68), with value y, direct dependencies PELION_SOURCES (value: y), and select condition PELION_SOURCES (value: y)
```

Which happens because the nrf_security symbols that has been selected are sticky, due to them having a prompt.

This error furthermore results in the inability to re-run `menuconfig`.

pdunaj replied:
I see but this is a generic problem of kconfig. If something is automagically selected it will stay selected even if you disable the top level thing. There is nothing you could do about it.
The `PELION_NRF_SECURITY` is a convenience symbol to select symbols that Pelion would like to use.

btw - it still has OpenThread text that went unnoticed. Will remove it.

tejlmand replied:
> There is nothing you could do about it.

yes, it all depends on how you structure your own Kconfig.
Users are expecting to use menuconfig in NCS, so we should not break such a basic feature.

pdunaj replied:
next fixed. I don't think this is a bug but a limitation of a config system. there is similar case in openthread as this file was based on it

pdunaj replied:
> Users are expecting to use menuconfig in NCS, so we should not break such a basic feature.

But there is nothing that can be done here. I cannot make both options `depend on` against each other as kconfig find this as a cyclic dependency error.
The same problem is with openthread convenience symbols - if you do the same trick menuconfig fails.

Note that using menuconfig in such cases is very limited as such "convenience symbols" often enable multiple others that will not be deselected. This is just the kconfig limitation.

We could remove this but in such case there will be no possibility to set this symbols on common (lib) level as you can only do that in kconfig.

pdunaj commented:
Fixed new checkpatch issues (@rlubos , @carlescufi  - checkpatch either does not run against c++ files or has different rules for them).
Fixed string in kconfig.
Rebased.

pdunaj commented:
One more checkpatch issue fixed.

emob-nordic commented on applications/pelion_client/src/events/Kconfig:
```diff
@@ -0,0 +1,28 @@
+
+comment "Initially logged events"
+
+config PELION_INIT_LOG_NET_STATE_EVENT
+	bool "Log net state event"
+	depends on LOG
+	default y
+
+config PELION_INIT_LOG_PELION_STATE_EVENT
+	bool "Log Pelion state event"
+	depends on LOG
+	default y
+
+config PELION_INIT_LOG_PELION_CREATE_OBJECTS_EVENT
```
`s/_EVENT/_EVENTS`
`s/state event"/state events"`
It's done similarly in the CAF, it would be consistent.

pdunaj replied:
why events? there is only one event?

pdunaj replied:
I see - we always use this form.

pdunaj replied:
Fixed

emob-nordic commented on applications/pelion_client/src/events/Kconfig.keep_alive_event:
```diff
@@ -0,0 +1,18 @@
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+config PELION_CLIENT_KEEP_ALIVE_EVENT_ENABLE
+	bool "Keep alive event"
```
``` suggestion
config PELION_CLIENT_KEEP_ALIVE_EVENTS
	bool "Enable keep alive events"
```

pdunaj replied:
Fixed

emob-nordic commented on applications/pelion_client/src/events/Kconfig.keep_alive_event:
```diff
@@ -0,0 +1,18 @@
+
+if PELION_CLIENT_KEEP_ALIVE_EVENT_ENABLE
+
+config PELION_CLIENT_INIT_LOG_KEEP_ALIVE_EVENT
+	bool "Log power manager keep alive event"
```
``` suggestion
config PELION_CLIENT_INIT_LOG_KEEP_ALIVE_EVENTS
	bool "Log power manager keep alive events"
```

pdunaj replied:
Fixed

emob-nordic commented on applications/pelion_client/src/events/Kconfig.keep_alive_event:
```diff
@@ -0,0 +1,18 @@
+config PELION_CLIENT_INIT_LOG_KEEP_ALIVE_EVENT
+	bool "Log power manager keep alive event"
+	help
+	  Log the keep alive event.
```
Help is not helpful if it says more or less same as string descriptor above, You can remove this.

pdunaj replied:
True but I will keep them as we have them for most things in this form.

emob-nordic commented on applications/pelion_client/src/modules/pelion_fcc.cpp:
```diff
@@ -0,0 +1,101 @@
+	}
+
+	if (reset_storage) {
+		LOG_WRN("Storage reset requested");
```
Not LOG_INF?

pdunaj replied:
This is important - let's keep it visible

emob-nordic commented on applications/pelion_client/src/modules/power_manager.c:
```diff
@@ -0,0 +1,254 @@
+	}
+
+	if (is_power_down_event(eh)) {
+		switch (power_state) {
```
Can You put handling of this event to separate function? `handle_power_down_event` ? Similarly for others?

pdunaj replied:
Can do but this is migration from repo to repo, not a code cleanup. I don't want to make too much changes here.
Once we get it in, have CI execute against the changes, we can make beautifications.

emob-nordic commented on lib/pelion/Kconfig:
```diff
@@ -0,0 +1,113 @@
+#
+# Copyright (c) 2020 Nordic Semiconductor ASA
```
2021, please check other places.

pdunaj replied:
Fixed

emob-nordic reviewed (changes requested):
Minor things.

pdunaj commented:
added changes from @tejlmand review.
One additional checkpatch issue due to symbol that will be introduced in subsequent change in zephyr
```
warning: undefined symbol MBEDTLS_CFG_FILE_PROMPT_OFF:
```

pdunaj commented:
LTE version works. Checking OT now.

tejlmand commented on lib/pelion/Kconfig:
```diff
@@ -0,0 +1,146 @@
+# The purpose of this file is to override and extend settings related to
+# building Pelion cloud client library.
+
+config MBEDTLS_PROMPTLESS
+	bool
+	default y if !CUSTOM_PELION_SECURITY
+
+config NORDIC_SECURITY_PROMPTLESS
+	bool
+	default y if !CUSTOM_PELION_SECURITY
+
```
Should be deleted, they have been moved to line L24-L31 to be inside the `PELION_SOURCES` safeguard.
```suggestion
```

pdunaj replied:
Removed

pdunaj commented:
removed mentioned lines in lib/pelion/Kconfig;
removed dependency workaround in apps/pelion/CMakeLists.txt
rebased on top of master

rlubos commented:
@pdunaj What about the documentation, currently there's no readme or whatsoever?

pdunaj commented:
rebased

pdunaj commented:
@rlubos 

>What about the documentation, currently there's no readme or whatsoever?

Is being worked on and will be added later on before the release.

pdunaj commented:
rebase due to failing CI
